### PR TITLE
client: support backoff mechanism for memberLoop

### PR DESCRIPTION
**Pull request by HuSharp**
Signed-off-by: husharp <jinhao.hu@pingcap.com>

### What problem does this PR solve?

Issue Number: ref #6556

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

### PR Summary
1. Add ready for resp
- Have goroutine `reconnectMemberLoop` call `updateMember` periodically. When calling `ScheduleCheckMemberChanged` channel, we need to wait for the goroutine to update members until ready or timeout.
2. Add backoff mechanism
- When waiting for the goroutine to update, the `expo` function can be used to backoff to sleep when an error is encountered.

### Reproduce Step

1. enable fail point, like gRPC is throttling, cannot read from etcd.
`curl -X PUT -d 'return(10)' http://tc-pd-1.tc-pd-peer.csn-simulator-big-cluster-vd62g.svc:2379/pd/api/v1/fail/github.com/tikv/pd/pkg/etcdutil/SlowEtcdKVGet`

2. simulate pd lost leader
`curl -X PUT -d 'return("2346857576170797299")' http://tc-pd-1.tc-pd-peer.csn-simulator-big-cluster-vd62g.svc:2379/pd/api/v1/fail/github.com/tikv/pd/server/exitCampaignLeader`

### Reproduce Result
Grpc request `GetMember` keeps high:
image

TiKV side show

image

### PR Effect

The Grpc `GetMember` call was reduced from 3.2k to 170, which is relative to the TiDB numbers and client requests for triaging checkLeader.

For 20 * tidb 3 * PD 50 * TiKV
170 = (50 * 3 / 3  / 3[TiKV side] + 20 * 2 [TiDB side]) * 3[PD Num]

And more tests are necessary to ensure that no further issues arise.

image

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/retry/backoff.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	totalSleep int
+
+	errors         []error
+	configs        []*Config
+	backoffSleepMS map[string]int
+	backoffTimes   map[string]int
+}
+
+// NewBackoffer (Deprecated) creates a Backoffer with maximum sleep time(in ms).
+func NewBackoffer(ctx context.Context, maxSleep int) *Backoffer {
```
Comment (nolouch) [inline client/retry/backoff.go:43]: There is no need for such a fully functional `backoff`. This is only used when reconnecting members, which is different from the scope of use of `client-go`.  make it simple easy to maintain and readable.

Comment (HuSharp) [reply client/retry/backoff.go]: fixed

File: client/base_client.go [hunk 2]
```diff
@@ -124,32 +132,133 @@ func (c *baseClient) initRetry(f func() error) error {
...
+	case c.successReConnect <- struct{}{}:
+	default:
+	}
+}
+
 // ScheduleCheckLeader is used to check leader.
 func (c *baseClient) ScheduleCheckLeader() {
 	select {
 	case c.checkLeaderCh <- struct{}{}:
+		if err := c.waitForReady(); err != nil {
```
[the remark this answers is not in the document]
Comment (HuSharp) [reply client/base_client.go]: removed

**Comment by nolouch** [comment]
plz base master.

File: client/base_client.go [hunk 3]
```diff
@@ -130,18 +131,24 @@ func (c *baseClient) memberLoop() {
...
+	defer ticker.Stop()
+
+	bo := retry.InitialBackOffer(100*time.Millisecond, updateMemberTimeout)
 	for {
 		select {
 		case <-c.checkLeaderCh:
-		case <-time.After(memberUpdateInterval):
+		case <-ticker.C:
 		case <-ctx.Done():
+			log.Info("[pd.memberLoop] exit memberLoop")
```
Comment (rleungx) [inline client/base_client.go:143]:
```suggestion
			log.Info("[pd] exit member loop")
```

File: client/retry/backoff.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	initialBackoff time.Duration
+}
+
+// WithBackoff is a helper function to add backoff.
+func WithBackoff(
+	ctx context.Context,
+	fn func() error,
+	bo *BackOffer,
+) error {
+	var allErrors error
```
[the remark this answers is not in the document]
Comment (HuSharp) [reply client/retry/backoff.go]: Oops, redundant line from previous implementation...

File: client/base_client.go [hunk 5]
```diff
@@ -130,18 +131,23 @@ func (c *baseClient) memberLoop() {
...
+	defer ticker.Stop()
+
+	bo := retry.InitialBackOffer(100*time.Millisecond, updateMemberTimeout)
 	for {
 		select {
 		case <-c.checkLeaderCh:
-		case <-time.After(memberUpdateInterval):
+		case <-ticker.C:
 		case <-ctx.Done():
+			log.Info("[pd] exit member loop")
```
Comment (JmPotato) [inline client/base_client.go:143]:
```suggestion
			log.Info("[pd] exit member loop due to context cancel")
```

File: client/retry/backoff.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+		case <-time.After(bo.NextBackoff()):
+			failpoint.Inject("backOffExecute", func() {
+				testBackOffExecuteFlag = true
+			})
+		}
+	} else {
+		bo.ResetBackoff()
+		return nil
+	}
+	return err
```
Comment (JmPotato) [inline client/retry/backoff.go:51]:
```suggestion
	}
	bo.ResetBackoff()
	return nil
```

Comment (HuSharp) [reply client/retry/backoff.go]: sry, remove the redundant line from the previous implementation with a lazy fix :(

**Review by JmPotato — APPROVED** [review]

File: client/retry/backoff.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	return backoff
+}
+
+// ResetBackoff reset the backoff to initial state.
+func (rs *BackOffer) ResetBackoff() {
+	rs.nextBackoff = rs.initialBackoff
+}
+
+// Only used for test.
+var testBackOffExecuteFlag = false
```
Comment (nolouch) [inline client/retry/backoff.go:82]: can it be removed?

File: client/base_client.go [hunk 8]
```diff
@@ -130,18 +131,23 @@ func (c *baseClient) memberLoop() {
 	ctx, cancel := context.WithCancel(c.ctx)
 	defer cancel()
 
+	ticker := time.NewTicker(memberUpdateInterval)
+	defer ticker.Stop()
+
+	bo := retry.InitialBackOffer(100*time.Millisecond, updateMemberTimeout)
```
Comment (nolouch) [inline client/base_client.go:137]: use const for `100 Millisecond`

File: client/retry/backoff.go (hunk 7 above)
Comment (HuSharp) [reply client/retry/backoff.go:82]: want to make sure back off executed

File: tests/pdctl/scheduler/scheduler_test.go [hunk 9]
```diff
@@ -397,22 +397,20 @@ func TestScheduler(t *testing.T) {
...
+	re.Equal(conf1["write-peer-priorities"], []interface{}{"byte", "key"})
 
 	// test remove and add
-	mustExec([]string{"-u", pdAddr, "scheduler", "remove", "balance-hot-region-scheduler"}, nil)
-	mustExec([]string{"-u", pdAddr, "scheduler", "add", "balance-hot-region-scheduler"}, nil)
-	re.Equal(expected1, conf1)
+	echo = mustExec([]string{"-u", pdAddr, "scheduler", "remove", "balance-hot-region-scheduler"}, nil)
+	re.Contains(echo, "Success")
+	echo = mustExec([]string{"-u", pdAddr, "scheduler", "add", "balance-hot-region-scheduler"}, nil)
+	re.Contains(echo, "Success")
```
Comment (HuSharp) [inline tests/pdctl/scheduler/scheduler_test.go:413]: cp https://github.com/tikv/pd/pull/5847 to make TestScheduler stable

**Review by nolouch — APPROVED** [review]

**Pull request merged**